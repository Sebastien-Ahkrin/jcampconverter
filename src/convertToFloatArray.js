export default function convertToFloatArray(stringArray) {
  let floatArray = [];
  for (let i = 0; i < stringArray.length; i++) {
    floatArray.push(parseFloat(stringArray[i]));
  }
  return floatArray;
}
