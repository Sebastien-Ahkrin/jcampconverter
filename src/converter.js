'use strict';

class JcampConverter {
    constructor(input, options) {

    }
}

module.exports = JcampConverter;