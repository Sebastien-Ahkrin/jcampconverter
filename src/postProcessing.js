import add2D from './2d/add2D';
import { complexChromatogram } from './complexChromatogram';
import simpleChromatogram from './simpleChromatogram';
import profiling from './profiling';

export default function postProcessing(entriesFlat, result, options) {
  for (let entry of entriesFlat) {
    if (Object.keys(entry.ntuples).length > 0) {
      let newNtuples = [];
      let keys = Object.keys(entry.ntuples);
      for (let i = 0; i < keys.length; i++) {
        let key = keys[i];
        let values = entry.ntuples[key];
        for (let j = 0; j < values.length; j++) {
          if (!newNtuples[j]) newNtuples[j] = {};
          newNtuples[j][key] = values[j];
        }
      }
      entry.ntuples = newNtuples;
    }

    if (entry.twoD && options.wantXY) {
      add2D(entry, options);

      profiling(result, 'Finished countour plot calculation', options);

      if (!options.keepSpectra) {
        delete entry.spectra;
      }
    }

    if (options.chromatogram) {
      options.xy = true;
    }

    if (options.xy && options.wantXY) {
      // the spectraData should not be a oneD array but an object with x and y
      if (entry.spectra && entry.spectra.length > 0) {
        for (let spectrum of entry.spectra) {
          if (spectrum.data) {
            let data = spectrum.data;
            let newData = {
              x: new Array(data.length / 2),
              y: new Array(data.length / 2),
            };
            for (let k = 0; k < data.length; k = k + 2) {
              newData.x[k / 2] = data[k];
              newData.y[k / 2] = data[k + 1];
            }
            spectrum.data = newData;
          }
        }
      }
    }

    // maybe it is a GC (HPLC) / MS. In this case we add a new format
    if (options.chromatogram) {
      if (entry.spectra.length > 1) {
        complexChromatogram(entry);
      } else {
        simpleChromatogram(entry);
      }
      profiling(result, 'Finished chromatogram calculation', options);
    }
  }
}
