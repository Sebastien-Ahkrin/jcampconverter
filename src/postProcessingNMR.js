import { gyromagneticRatio } from 'nmr-processing';

export default function postProcessingNMR(entriesFlat) {
  // specific NMR functions
  let observeFrequency = 0;
  let shiftOffsetVal = 0;

  for (let entry of entriesFlat) {
    for (let spectrum of entry.spectra) {
      if (entry.ntuples && entry.ntuples.symbol) {
        if (!observeFrequency && spectrum.observeFrequency) {
          observeFrequency = spectrum.observeFrequency;
        }
        if (!shiftOffsetVal && spectrum.shiftOffsetVal) {
          shiftOffsetVal = spectrum.shiftOffsetVal;
        }
      } else {
        observeFrequency = spectrum.observeFrequency;
        shiftOffsetVal = spectrum.shiftOffsetVal;
      }

      if (observeFrequency) {
        if (spectrum.xUnits && spectrum.xUnits.toUpperCase().includes('HZ')) {
          spectrum.xUnits = 'PPM';
          spectrum.xFactor = spectrum.xFactor / observeFrequency;
          spectrum.firstX = spectrum.firstX / observeFrequency;
          spectrum.lastX = spectrum.lastX / observeFrequency;
          spectrum.deltaX = spectrum.deltaX / observeFrequency;
          for (let i = 0; i < spectrum.data.x.length; i++) {
            spectrum.data.x[i] /= observeFrequency;
          }
        }
      }
      if (shiftOffsetVal) {
        let shift = spectrum.firstX - shiftOffsetVal;
        spectrum.firstX = spectrum.firstX - shift;
        spectrum.lastX = spectrum.lastX - shift;
        for (let i = 0; i < spectrum.data.x.length; i++) {
          spectrum.data.x[i] -= shift;
        }
      }

      if (
        observeFrequency &&
        entry.ntuples &&
        entry.ntuples.symbol &&
        entry.ntuples.nucleus
      ) {
        let unit = '';
        let pageSymbolIndex = entry.ntuples.symbol.indexOf(spectrum.pageSymbol);
        if (entry.ntuples.units && entry.ntuples.units[pageSymbolIndex]) {
          unit = entry.ntuples.units[pageSymbolIndex];
        }
        if (unit !== 'PPM') {
          if (pageSymbolIndex !== 0) {
            throw Error('Not sure about this ntuples format');
          }

          let ratio0 = gyromagneticRatio[entry.ntuples.nucleus[0]];
          let ratio1 = gyromagneticRatio[entry.ntuples.nucleus[1]];
          if (!ratio0 || !ratio1) {
            throw Error('Problem with determination of gyromagnetic ratio');
          }
          let ratio = (ratio0 / ratio1) * observeFrequency;
          spectrum.pageValue /= ratio;
        }
      }
    }
  }
}
