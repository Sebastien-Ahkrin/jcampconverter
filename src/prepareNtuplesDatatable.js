export default function prepareNtuplesDatatable(currentEntry, spectrum, kind) {
  let xIndex = -1;
  let yIndex = -1;
  let firstVariable = '';
  let secondVariable = '';
  if (kind.indexOf('++') > 0) {
    firstVariable = kind.replace(/.*\(([a-zA-Z0-9]+)\+\+.*/, '$1');
    secondVariable = kind.replace(/.*\.\.([a-zA-Z0-9]+).*/, '$1');
  } else {
    kind = kind.replace(/[^a-zA-Z]/g, '');
    firstVariable = kind.charAt(0);
    secondVariable = kind.charAt(1);
    spectrum.variables = {};
    for (let symbol of kind) {
      let index = currentEntry.ntuples.symbol.indexOf(symbol);
      if (index === -1) throw Error(`Symbol undefined: ${symbol}`);
      let lowercaseSymbol = symbol.toLowerCase();
      spectrum.variables[lowercaseSymbol] = {};
      for (let key in currentEntry.ntuples) {
        if (currentEntry.ntuples[key][index]) {
          spectrum.variables[lowercaseSymbol][key] =
            currentEntry.ntuples[key][index];
        }
      }
    }
  }
  xIndex = currentEntry.ntuples.symbol.indexOf(firstVariable);
  yIndex = currentEntry.ntuples.symbol.indexOf(secondVariable);

  if (xIndex === -1) xIndex = 0;
  if (yIndex === -1) yIndex = 0;

  if (currentEntry.ntuples.first) {
    if (currentEntry.ntuples.first.length > xIndex) {
      spectrum.firstX = currentEntry.ntuples.first[xIndex];
    }
    if (currentEntry.ntuples.first.length > yIndex) {
      spectrum.firstY = currentEntry.ntuples.first[yIndex];
    }
  }
  if (currentEntry.ntuples.last) {
    if (currentEntry.ntuples.last.length > xIndex) {
      spectrum.lastX = currentEntry.ntuples.last[xIndex];
    }
    if (currentEntry.ntuples.last.length > yIndex) {
      spectrum.lastY = currentEntry.ntuples.last[yIndex];
    }
  }
  if (
    currentEntry.ntuples.vardim &&
    currentEntry.ntuples.vardim.length > xIndex
  ) {
    spectrum.nbPoints = currentEntry.ntuples.vardim[xIndex];
  }
  if (currentEntry.ntuples.factor) {
    if (currentEntry.ntuples.factor.length > xIndex) {
      spectrum.xFactor = currentEntry.ntuples.factor[xIndex];
    }
    if (currentEntry.ntuples.factor.length > yIndex) {
      spectrum.yFactor = currentEntry.ntuples.factor[yIndex];
    }
  }
  if (currentEntry.ntuples.units) {
    if (currentEntry.ntuples.units.length > xIndex) {
      spectrum.xUnit = currentEntry.ntuples.units[xIndex];
    }
    if (currentEntry.ntuples.units.length > yIndex) {
      spectrum.yUnit = currentEntry.ntuples.units[yIndex];
    }
  }
}
