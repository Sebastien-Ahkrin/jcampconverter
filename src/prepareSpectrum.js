export default function prepareSpectrum(spectrum) {
  if (!spectrum.xFactor) spectrum.xFactor = 1;
  if (!spectrum.yFactor) spectrum.yFactor = 1;
  if (spectrum.observeFrequency) {
    if (spectrum.xUnit && spectrum.xUnit.toUpperCase() === 'HZ') {
      spectrum.xUnit = 'PPM';
      spectrum.xFactor = spectrum.xFactor / spectrum.observeFrequency;
      spectrum.firstX = spectrum.firstX / spectrum.observeFrequency;
      spectrum.lastX = spectrum.lastX / spectrum.observeFrequency;
      spectrum.deltaX = spectrum.deltaX / spectrum.observeFrequency;
    }
  }
  if (spectrum.shiftOffsetVal) {
    let shift = spectrum.firstX - spectrum.shiftOffsetVal;
    spectrum.firstX = spectrum.firstX - shift;
    spectrum.lastX = spectrum.lastX - shift;
  }
}
