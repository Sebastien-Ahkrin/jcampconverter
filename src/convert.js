import { isMSField, convertMSFieldToLabel } from './complexChromatogram';
import convertToFloatArray from './convertToFloatArray';
import fastParseXYData from './parse/fastParseXYData';
import parsePeakTable from './parse/parsePeakTable';
import parseXYA from './parse/parseXYA';
import postProcessing from './postProcessing';
import prepareSpectrum from './prepareSpectrum';
import profiling from './profiling';

// the following RegExp can only be used for XYdata, some peakTables have values with a "E-5" ...
const ntuplesSeparator = /[, \t]+/;

class Spectrum {}

const defaultOptions = {
  keepRecordsRegExp: /^$/,
  canonicDataLabels: true,
  dynamicTyping: false,
  xy: true,
  withoutXY: false,
  chromatogram: false,
  keepSpectra: false,
  noContour: false,
  nbContourLevels: 7,
  noiseMultiplier: 5,
  profiling: false,
};

export default function convert(jcamp, options) {
  options = Object.assign({}, defaultOptions, options);
  options.wantXY = !options.withoutXY;
  options.start = Date.now();

  let entriesFlat = [];

  let result = {
    profiling: options.profiling ? [] : false,
    logs: [],
    entries: [],
  };

  let tmpResult = { children: [] };
  let currentEntry = tmpResult;
  let parentsStack = [];

  let spectrum = new Spectrum();

  if (typeof jcamp !== 'string') {
    throw new TypeError('the JCAMP should be a string');
  }

  profiling(result, 'Before split to LDRS', options);

  let ldrs = jcamp.split(/[\r\n]+##/);

  profiling(result, 'Split to LDRS', options);

  if (ldrs[0]) ldrs[0] = ldrs[0].replace(/^[\r\n ]*##/, '');

  for (let ldr of ldrs) {
    // This is a new LDR
    let position = ldr.indexOf('=');
    let dataLabel = position > 0 ? ldr.substring(0, position) : ldr;
    let dataValue = position > 0 ? ldr.substring(position + 1).trim() : '';

    let canonicDataLabel = dataLabel.replace(/[_ -]/g, '').toUpperCase();

    if (canonicDataLabel === 'DATATABLE') {
      let endLine = dataValue.indexOf('\n');
      if (endLine === -1) endLine = dataValue.indexOf('\r');
      if (endLine > 0) {
        let xIndex = -1;
        let yIndex = -1;
        // ##DATA TABLE= (X++(I..I)), XYDATA
        // We need to find the variables

        let infos = dataValue.substring(0, endLine).split(/[ ,;\t]+/);
        if (infos[0].indexOf('++') > 0) {
          let firstVariable = infos[0].replace(
            /.*\(([a-zA-Z0-9]+)\+\+.*/,
            '$1',
          );
          let secondVariable = infos[0].replace(/.*\.\.([a-zA-Z0-9]+).*/, '$1');
          xIndex = currentEntry.ntuples.symbol.indexOf(firstVariable);
          yIndex = currentEntry.ntuples.symbol.indexOf(secondVariable);
        }

        if (xIndex === -1) xIndex = 0;
        if (yIndex === -1) yIndex = 0;

        if (currentEntry.ntuples.first) {
          if (currentEntry.ntuples.first.length > xIndex) {
            spectrum.firstX = currentEntry.ntuples.first[xIndex];
          }
          if (currentEntry.ntuples.first.length > yIndex) {
            spectrum.firstY = currentEntry.ntuples.first[yIndex];
          }
        }
        if (currentEntry.ntuples.last) {
          if (currentEntry.ntuples.last.length > xIndex) {
            spectrum.lastX = currentEntry.ntuples.last[xIndex];
          }
          if (currentEntry.ntuples.last.length > yIndex) {
            spectrum.lastY = currentEntry.ntuples.last[yIndex];
          }
        }
        if (
          currentEntry.ntuples.vardim &&
          currentEntry.ntuples.vardim.length > xIndex
        ) {
          spectrum.nbPoints = currentEntry.ntuples.vardim[xIndex];
        }
        if (currentEntry.ntuples.factor) {
          if (currentEntry.ntuples.factor.length > xIndex) {
            spectrum.xFactor = currentEntry.ntuples.factor[xIndex];
          }
          if (currentEntry.ntuples.factor.length > yIndex) {
            spectrum.yFactor = currentEntry.ntuples.factor[yIndex];
          }
        }
        if (currentEntry.ntuples.units) {
          if (currentEntry.ntuples.units.length > xIndex) {
            spectrum.xUnit = currentEntry.ntuples.units[xIndex];
          }
          if (currentEntry.ntuples.units.length > yIndex) {
            spectrum.yUnit = currentEntry.ntuples.units[yIndex];
          }
        }
        spectrum.datatable = infos[0];
        if (infos[1] && infos[1].indexOf('PEAKS') > -1) {
          canonicDataLabel = 'PEAKTABLE';
        } else if (
          infos[1] &&
          (infos[1].indexOf('XYDATA') || infos[0].indexOf('++') > 0)
        ) {
          canonicDataLabel = 'XYDATA';
          spectrum.deltaX =
            (spectrum.lastX - spectrum.firstX) / (spectrum.nbPoints - 1);
        }
      }
    }

    if (canonicDataLabel === 'XYDATA') {
      if (options.wantXY) {
        prepareSpectrum(spectrum);
        // well apparently we should still consider it is a PEAK TABLE if there are no '++' after
        if (dataValue.match(/.*\+\+.*/)) {
          // ex: (X++(Y..Y))
          if (!spectrum.deltaX) {
            spectrum.deltaX =
              (spectrum.lastX - spectrum.firstX) / (spectrum.nbPoints - 1);
          }
          fastParseXYData(spectrum, dataValue, result);
        } else {
          parsePeakTable(spectrum, dataValue, result);
        }
        currentEntry.spectra.push(spectrum);
        spectrum = new Spectrum();
      }
      continue;
    } else if (canonicDataLabel === 'PEAKTABLE') {
      if (options.wantXY) {
        prepareSpectrum(spectrum);
        parsePeakTable(spectrum, dataValue, result);
        currentEntry.spectra.push(spectrum);
        spectrum = new Spectrum();
      }
      continue;
    }
    if (canonicDataLabel === 'PEAKASSIGNMENTS') {
      if (options.wantXY) {
        if (dataValue.match(/.*(XYA).*/)) {
          // ex: (XYA)
          parseXYA(spectrum, dataValue);
        }
        currentEntry.spectra.push(spectrum);
        spectrum = new Spectrum();
      }
      continue;
    }

    if (canonicDataLabel === 'TITLE') {
      let parentEntry = currentEntry;
      if (!parentEntry.children) {
        parentEntry.children = [];
      }
      currentEntry = {
        spectra: [],
        ntuples: {},
        info: {},
      };
      parentEntry.children.push(currentEntry);
      parentsStack.push(parentEntry);
      entriesFlat.push(currentEntry);
      currentEntry.title = dataValue;
    } else if (canonicDataLabel === 'DATATYPE') {
      currentEntry.dataType = dataValue;
      if (dataValue.indexOf('nD') > -1) {
        currentEntry.twoD = true;
      }
    } else if (canonicDataLabel === 'NTUPLES') {
      if (dataValue.indexOf('nD') > -1) {
        currentEntry.twoD = true;
      }
    } else if (canonicDataLabel === 'DATACLASS') {
      currentEntry.dataClass = dataValue;
    } else if (canonicDataLabel === 'XUNITS') {
      spectrum.xUnit = dataValue;
    } else if (canonicDataLabel === 'YUNITS') {
      spectrum.yUnit = dataValue;
    } else if (canonicDataLabel === 'FIRSTX') {
      spectrum.firstX = parseFloat(dataValue);
    } else if (canonicDataLabel === 'LASTX') {
      spectrum.lastX = parseFloat(dataValue);
    } else if (canonicDataLabel === 'FIRSTY') {
      spectrum.firstY = parseFloat(dataValue);
    } else if (canonicDataLabel === 'LASTY') {
      spectrum.lastY = parseFloat(dataValue);
    } else if (canonicDataLabel === 'NPOINTS') {
      spectrum.nbPoints = parseFloat(dataValue);
    } else if (canonicDataLabel === 'XFACTOR') {
      spectrum.xFactor = parseFloat(dataValue);
    } else if (canonicDataLabel === 'YFACTOR') {
      spectrum.yFactor = parseFloat(dataValue);
    } else if (canonicDataLabel === 'MAXX') {
      spectrum.maxX = parseFloat(dataValue);
    } else if (canonicDataLabel === 'MINX') {
      spectrum.minX = parseFloat(dataValue);
    } else if (canonicDataLabel === 'MAXY') {
      spectrum.maxY = parseFloat(dataValue);
    } else if (canonicDataLabel === 'MINY') {
      spectrum.minY = parseFloat(dataValue);
    } else if (canonicDataLabel === 'DELTAX') {
      spectrum.deltaX = parseFloat(dataValue);
    } else if (
      canonicDataLabel === '.OBSERVEFREQUENCY' ||
      canonicDataLabel === '$SFO1'
    ) {
      if (!spectrum.observeFrequency) {
        spectrum.observeFrequency = parseFloat(dataValue);
      }
    } else if (canonicDataLabel === '.OBSERVENUCLEUS') {
      if (!spectrum.xType) {
        currentEntry.xType = dataValue.replace(/[^a-zA-Z0-9]/g, '');
      }
    } else if (canonicDataLabel === '$SFO2') {
      if (!currentEntry.indirectFrequency) {
        currentEntry.indirectFrequency = parseFloat(dataValue);
      }
    } else if (canonicDataLabel === '$OFFSET') {
      // OFFSET for Bruker spectra
      currentEntry.shiftOffsetNum = 0;
      if (!spectrum.shiftOffsetVal) {
        spectrum.shiftOffsetVal = parseFloat(dataValue);
      }
    } else if (canonicDataLabel === '$REFERENCEPOINT') {
      // OFFSET for Varian spectra
      // if we activate this part it does not work for ACD specmanager
      //         } else if (canonicDataLabel=='.SHIFTREFERENCE') {   // OFFSET FOR Bruker Spectra
      //                 var parts = dataValue.split(/ *, */);
      //                 currentEntry.shiftOffsetNum = parseInt(parts[2].trim());
      //                 spectrum.shiftOffsetVal = parseFloat(parts[3].trim());
    } else if (canonicDataLabel === 'VARNAME') {
      currentEntry.ntuples.varname = dataValue.split(ntuplesSeparator);
    } else if (canonicDataLabel === 'SYMBOL') {
      currentEntry.ntuples.symbol = dataValue.split(ntuplesSeparator);
    } else if (canonicDataLabel === 'VARTYPE') {
      currentEntry.ntuples.vartype = dataValue.split(ntuplesSeparator);
    } else if (canonicDataLabel === 'VARFORM') {
      currentEntry.ntuples.varform = dataValue.split(ntuplesSeparator);
    } else if (canonicDataLabel === 'VARDIM') {
      currentEntry.ntuples.vardim = convertToFloatArray(
        dataValue.split(ntuplesSeparator),
      );
    } else if (canonicDataLabel === 'UNITS') {
      currentEntry.ntuples.units = dataValue.split(ntuplesSeparator);
    } else if (canonicDataLabel === 'FACTOR') {
      currentEntry.ntuples.factor = convertToFloatArray(
        dataValue.split(ntuplesSeparator),
      );
    } else if (canonicDataLabel === 'FIRST') {
      currentEntry.ntuples.first = convertToFloatArray(
        dataValue.split(ntuplesSeparator),
      );
    } else if (canonicDataLabel === 'LAST') {
      currentEntry.ntuples.last = convertToFloatArray(
        dataValue.split(ntuplesSeparator),
      );
    } else if (canonicDataLabel === 'MIN') {
      currentEntry.ntuples.min = convertToFloatArray(
        dataValue.split(ntuplesSeparator),
      );
    } else if (canonicDataLabel === 'MAX') {
      currentEntry.ntuples.max = convertToFloatArray(
        dataValue.split(ntuplesSeparator),
      );
    } else if (canonicDataLabel === '.NUCLEUS') {
      if (currentEntry.twoD) {
        currentEntry.yType = dataValue.split(ntuplesSeparator)[0];
      }
    } else if (canonicDataLabel === 'PAGE') {
      spectrum.page = dataValue.trim();
      spectrum.pageValue = parseFloat(dataValue.replace(/^.*=/, ''));
      spectrum.pageSymbol = spectrum.page.replace(/[=].*/, '');
      let pageSymbolIndex = currentEntry.ntuples.symbol.indexOf(
        spectrum.pageSymbol,
      );
      let unit = '';
      if (
        currentEntry.ntuples.units &&
        currentEntry.ntuples.units[pageSymbolIndex]
      ) {
        unit = currentEntry.ntuples.units[pageSymbolIndex];
      }
      if (currentEntry.indirectFrequency && unit !== 'PPM') {
        spectrum.pageValue /= currentEntry.indirectFrequency;
      }
    } else if (canonicDataLabel === 'RETENTIONTIME') {
      spectrum.pageValue = parseFloat(dataValue);
    } else if (isMSField(canonicDataLabel)) {
      spectrum[convertMSFieldToLabel(canonicDataLabel)] = dataValue;
    } else if (canonicDataLabel === 'SAMPLEDESCRIPTION') {
      spectrum.sampleDescription = dataValue;
    } else if (canonicDataLabel === 'END') {
      currentEntry = parentsStack.pop();
    }

    if (
      currentEntry &&
      currentEntry.info &&
      canonicDataLabel.match(options.keepRecordsRegExp)
    ) {
      let label = options.canonicDataLabels ? canonicDataLabel : dataLabel;
      let value = dataValue.trim();
      if (options.dynamicTyping && !isNaN(value)) {
        value = Number(value);
      }
      if (currentEntry.info[label]) {
        if (!Array.isArray(currentEntry.info[label])) {
          currentEntry.info[label] = [currentEntry.info[label]];
        }
        currentEntry.info[label].push(value);
      } else {
        currentEntry.info[label] = value;
      }
    }
  }

  profiling(result, 'Finished parsing', options);

  postProcessing(entriesFlat, result, options);

  profiling(result, 'Total time', options);

  /*
  if (result.children && result.children.length>0) {
    result = { ...result, ...result.children[0] };
  }
  */
  result.entries = tmpResult.children;
  result.flatten = entriesFlat;

  return result;
}
