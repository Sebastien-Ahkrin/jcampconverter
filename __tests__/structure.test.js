import { readFileSync } from 'fs';
import { join } from 'path';

import { convert } from '../src';

describe('Test various jcamp structures', () => {
  it('simple spectrum', () => {
    let result = convert(
      readFileSync(join(__dirname, '/data/structure/normal.jdx'), 'utf8'),
    );
    expect(result.spectra).toHaveLength(1);
    expect(result.spectra[0].title).toBe('1 2 3');
    expect(result.spectra[0].dataType).toBe('type 1');
    expect(result.spectra[0].data).toStrictEqual({ x: [1, 2], y: [2, 3] });
  });

  it('two spectra', () => {
    let result = convert(
      readFileSync(join(__dirname, '/data/structure/two.jdx'), 'utf8'),
    );
    expect(result.spectra).toHaveLength(2);
    expect(result.spectra[0].title).toBe('1 2 3');
    expect(result.spectra[0].dataType).toBe('type 1');
    expect(result.spectra[0].data).toStrictEqual({ x: [1, 2], y: [2, 3] });
    expect(result.spectra[1].title).toBe('3 4 5');
    expect(result.spectra[1].dataType).toBe('type 2');
    expect(result.spectra[1].data).toStrictEqual({ x: [3, 4], y: [4, 5] });
  });

  it('reim', () => {
    let result = convert(
      readFileSync(join(__dirname, '/data/structure/reim.jdx'), 'utf8'),
    );
    expect(result.spectra).toHaveLength(2);
    expect(result.spectra[0].data).toStrictEqual({
      x: [1, 2, 3],
      y: [2, 3, 4],
    });
    expect(result.spectra[1].data).toStrictEqual({
      x: [1, 2, 3],
      y: [3, 4, 5],
    });
  });

  it.skip('ntuples', () => {
    let result = convert(
      readFileSync(join(__dirname, '/data/structure/ntuples.jdx'), 'utf8'),
      { noContour: true, keepSpectra: true },
    );
    expect(result.spectra).toHaveLength(2);
    expect(result.spectra[0].data).toStrictEqual({
      x: [1, 2, 3],
      y: [2, 3, 4],
    });
    expect(result.spectra[1].data).toStrictEqual({
      x: [1, 2, 3],
      y: [3, 4, 5],
    });
  });

  it('tree', () => {
    let result = convert(
      readFileSync(join(__dirname, '/data/structure/tree.jdx'), 'utf8'),
    );
    // console.log(result);
  });
});
