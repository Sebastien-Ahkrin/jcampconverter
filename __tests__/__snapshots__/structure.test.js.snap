// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test various jcamp structures assigned 1`] = `
Object {
  "entries": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "info": Object {
                "ATOMLIST": "$$ AN	AS	NH
	1	C	 
	2	C",
                "BONDLIST": "$$ AN1	AN2	BT
	1	2	D
	2	3	S",
                "TITLE": "Molecule",
              },
              "meta": Object {},
              "ntuples": Object {},
              "spectra": Array [],
              "title": "Molecule",
            },
            Object {
              "dataClass": "ASSIGNMENTS",
              "dataType": "NMR PEAK ASSIGNMENTS",
              "info": Object {
                "DATACLASS": "ASSIGNMENTS",
                "DATATYPE": "NMR PEAK ASSIGNMENTS",
                "TITLE": "Assignment",
              },
              "meta": Object {},
              "ntuples": Object {},
              "spectra": Array [
                Spectrum {},
              ],
              "title": "Assignment",
            },
          ],
          "dataType": "LINK 2",
          "info": Object {
            "DATATYPE": "LINK 2",
            "END": "",
            "TITLE": "title link 2",
          },
          "meta": Object {},
          "ntuples": Object {},
          "spectra": Array [],
          "title": "title link 2",
        },
        Object {
          "dataType": "type 1",
          "info": Object {
            "DATATYPE": "type 1",
            "NPOINTS": 2,
            "TITLE": "1 2 3",
          },
          "meta": Object {
            "meta info": "META",
          },
          "ntuples": Object {},
          "spectra": Array [
            Spectrum {
              "data": Object {
                "x": Array [
                  1,
                  2,
                ],
                "y": Array [
                  2,
                  3,
                ],
              },
              "isPeaktable": true,
              "nbPoints": 2,
              "xFactor": 1,
              "yFactor": 1,
            },
          ],
          "title": "1 2 3",
        },
        Object {
          "dataType": "type 2",
          "info": Object {
            "DATATYPE": "type 2",
            "TITLE": "peaks",
          },
          "meta": Object {},
          "ntuples": Object {},
          "spectra": Array [
            Spectrum {
              "data": Object {
                "x": Array [],
                "y": Array [],
              },
              "isPeaktable": true,
              "xFactor": 1,
              "yFactor": 1,
            },
          ],
          "title": "peaks",
        },
      ],
      "dataType": "LINK 1",
      "info": Object {
        "DATATYPE": "LINK 1",
        "END": "",
        "TITLE": "title link 1",
      },
      "meta": Object {},
      "ntuples": Object {},
      "spectra": Array [],
      "title": "title link 1",
    },
  ],
  "flatten": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "info": Object {
                "ATOMLIST": "$$ AN	AS	NH
	1	C	 
	2	C",
                "BONDLIST": "$$ AN1	AN2	BT
	1	2	D
	2	3	S",
                "TITLE": "Molecule",
              },
              "meta": Object {},
              "ntuples": Object {},
              "spectra": Array [],
              "title": "Molecule",
            },
            Object {
              "dataClass": "ASSIGNMENTS",
              "dataType": "NMR PEAK ASSIGNMENTS",
              "info": Object {
                "DATACLASS": "ASSIGNMENTS",
                "DATATYPE": "NMR PEAK ASSIGNMENTS",
                "TITLE": "Assignment",
              },
              "meta": Object {},
              "ntuples": Object {},
              "spectra": Array [
                Spectrum {},
              ],
              "title": "Assignment",
            },
          ],
          "dataType": "LINK 2",
          "info": Object {
            "DATATYPE": "LINK 2",
            "END": "",
            "TITLE": "title link 2",
          },
          "meta": Object {},
          "ntuples": Object {},
          "spectra": Array [],
          "title": "title link 2",
        },
        Object {
          "dataType": "type 1",
          "info": Object {
            "DATATYPE": "type 1",
            "NPOINTS": 2,
            "TITLE": "1 2 3",
          },
          "meta": Object {
            "meta info": "META",
          },
          "ntuples": Object {},
          "spectra": Array [
            Spectrum {
              "data": Object {
                "x": Array [
                  1,
                  2,
                ],
                "y": Array [
                  2,
                  3,
                ],
              },
              "isPeaktable": true,
              "nbPoints": 2,
              "xFactor": 1,
              "yFactor": 1,
            },
          ],
          "title": "1 2 3",
        },
        Object {
          "dataType": "type 2",
          "info": Object {
            "DATATYPE": "type 2",
            "TITLE": "peaks",
          },
          "meta": Object {},
          "ntuples": Object {},
          "spectra": Array [
            Spectrum {
              "data": Object {
                "x": Array [],
                "y": Array [],
              },
              "isPeaktable": true,
              "xFactor": 1,
              "yFactor": 1,
            },
          ],
          "title": "peaks",
        },
      ],
      "dataType": "LINK 1",
      "info": Object {
        "DATATYPE": "LINK 1",
        "END": "",
        "TITLE": "title link 1",
      },
      "meta": Object {},
      "ntuples": Object {},
      "spectra": Array [],
      "title": "title link 1",
    },
    Object {
      "children": Array [
        Object {
          "info": Object {
            "ATOMLIST": "$$ AN	AS	NH
	1	C	 
	2	C",
            "BONDLIST": "$$ AN1	AN2	BT
	1	2	D
	2	3	S",
            "TITLE": "Molecule",
          },
          "meta": Object {},
          "ntuples": Object {},
          "spectra": Array [],
          "title": "Molecule",
        },
        Object {
          "dataClass": "ASSIGNMENTS",
          "dataType": "NMR PEAK ASSIGNMENTS",
          "info": Object {
            "DATACLASS": "ASSIGNMENTS",
            "DATATYPE": "NMR PEAK ASSIGNMENTS",
            "TITLE": "Assignment",
          },
          "meta": Object {},
          "ntuples": Object {},
          "spectra": Array [
            Spectrum {},
          ],
          "title": "Assignment",
        },
      ],
      "dataType": "LINK 2",
      "info": Object {
        "DATATYPE": "LINK 2",
        "END": "",
        "TITLE": "title link 2",
      },
      "meta": Object {},
      "ntuples": Object {},
      "spectra": Array [],
      "title": "title link 2",
    },
    Object {
      "info": Object {
        "ATOMLIST": "$$ AN	AS	NH
	1	C	 
	2	C",
        "BONDLIST": "$$ AN1	AN2	BT
	1	2	D
	2	3	S",
        "TITLE": "Molecule",
      },
      "meta": Object {},
      "ntuples": Object {},
      "spectra": Array [],
      "title": "Molecule",
    },
    Object {
      "dataClass": "ASSIGNMENTS",
      "dataType": "NMR PEAK ASSIGNMENTS",
      "info": Object {
        "DATACLASS": "ASSIGNMENTS",
        "DATATYPE": "NMR PEAK ASSIGNMENTS",
        "TITLE": "Assignment",
      },
      "meta": Object {},
      "ntuples": Object {},
      "spectra": Array [
        Spectrum {},
      ],
      "title": "Assignment",
    },
    Object {
      "dataType": "type 1",
      "info": Object {
        "DATATYPE": "type 1",
        "NPOINTS": 2,
        "TITLE": "1 2 3",
      },
      "meta": Object {
        "meta info": "META",
      },
      "ntuples": Object {},
      "spectra": Array [
        Spectrum {
          "data": Object {
            "x": Array [
              1,
              2,
            ],
            "y": Array [
              2,
              3,
            ],
          },
          "isPeaktable": true,
          "nbPoints": 2,
          "xFactor": 1,
          "yFactor": 1,
        },
      ],
      "title": "1 2 3",
    },
    Object {
      "dataType": "type 2",
      "info": Object {
        "DATATYPE": "type 2",
        "TITLE": "peaks",
      },
      "meta": Object {},
      "ntuples": Object {},
      "spectra": Array [
        Spectrum {
          "data": Object {
            "x": Array [],
            "y": Array [],
          },
          "isPeaktable": true,
          "xFactor": 1,
          "yFactor": 1,
        },
      ],
      "title": "peaks",
    },
  ],
  "logs": Array [
    "Format error: 1,2,3",
    "Format error: 2,3,4",
    "Format error: 3,4,5",
  ],
  "profiling": false,
}
`;
